src/app/introduction/picture-puzzle.component.ts

```diff
@@ -0,0 +1,370 @@
+        if (this._currentPiece != null) {
+            this._stage.clearRect(this._currentPiece.xPos, this._currentPiece.yPos, this._pieceWidth, this._pieceHeight);
+            this._stage.save();
+            this._stage.globalAlpha = .9;
+            this._stage.drawImage(this._newImg, this._currentPiece.sx, this._currentPiece.sy, this._pieceWidth, this._pieceHeight, this._mouse.x - (this._pieceWidth / 2), this._mouse.y - (this._pieceHeight / 2), this._pieceWidth, this._pieceHeight);
+            this._stage.restore();
+            this._mousemoveListener = this._renderer.listen(this._canvas, 'mousemove', (event) => this.updatePuzzle(event));
+            this._mouseupListener = this._renderer.listen(this._canvas, 'mouseup', (event) => this.pieceDropped(event));
+        }
+    }
+
+    checkPieceClicked() {
+        let i, piece;
+        for (i = 0; i < this._pieces.length; i++) {
+            piece = this._pieces[i];
+            if (this._mouse.x < piece.xPos || this._mouse.x > (piece.xPos + this._pieceWidth) || this._mouse.y < piece.yPos || this._mouse.y > (piece.yPos + this._pieceHeight)) {
+                // Piece not at the right place
+            } else {
+                return piece;
+            }
+        }
+        return null;
+    }
+
+    updatePuzzle(e) {
+        this._currentDropPiece = null;
+        this.eventHandler(e);
+        this._stage.clearRect(0, 0, this._puzzleWidth, this._puzzleHeight);
+        let i, piece;
+        for (i = 0; i < this._pieces.length; i++) {
+            piece = this._pieces[i];
+            if (piece === this._currentPiece) {
+                continue;
+            }
+            this._stage.drawImage(this._newImg, piece.sx, piece.sy, this._pieceWidth, this._pieceHeight, piece.xPos, piece.yPos, this._pieceWidth, this._pieceHeight);
+            this._stage.strokeRect(piece.xPos, piece.yPos, this._pieceWidth, this._pieceHeight);
+            if (this._currentDropPiece == null) {
+                if (this._mouse.x < piece.xPos || this._mouse.x > (piece.xPos + this._pieceWidth) || this._mouse.y < piece.yPos || this._mouse.y > (piece.yPos + this._pieceHeight)) {
+                    // Game not over
+                } else {
+                    this._currentDropPiece = piece;
+                    this._stage.save();
+                    this._stage.globalAlpha = .4;
+                    this._stage.fillStyle = this.PUZZLE_HOVER_TINT;
+                    this._stage.fillRect(this._currentDropPiece.xPos, this._currentDropPiece.yPos, this._pieceWidth, this._pieceHeight);
+                    this._stage.restore();
+                }
+            }
+        }
+        this._stage.save();
+        this._stage.globalAlpha = .6;
+        this._stage.drawImage(this._newImg, this._currentPiece.sx, this._currentPiece.sy, this._pieceWidth, this._pieceHeight, this._mouse.x - (this._pieceWidth / 2), this._mouse.y - (this._pieceHeight / 2), this._pieceWidth, this._pieceHeight);
```

Comment:
i actually going to variables for ` this._mouse.x - (this._pieceWidth / 2)` & `this._mouse.y - (this._pieceHeight / 2)` to reduce the amount of calculation needed